module/zstd/zstd.c

```diff
@@ -0,0 +1,618 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016-2018, Klara Systems Inc. All rights reserved.
+ * Copyright (c) 2016-2018, Allan Jude. All rights reserved.
+ * Copyright (c) 2018-2019, Sebastian Gottschall. All rights reserved.
+ * Copyright (c) 2019, Michael Niewöhner. All rights reserved.
+ */
+
+#include <sys/param.h>
+#include <sys/sysmacros.h>
+#include <sys/zfs_context.h>
+#include <sys/zio_compress.h>
+#include <sys/spa.h>
+#include <sys/zstd/zstd.h>
+
+#define	ZSTD_STATIC_LINKING_ONLY
+#include <zstd.h>
+#include <zstd_errors.h>
+#include <error_private.h>
+
+/* For BSD compatibility */
+#define	__unused	__attribute__((unused))
+
+/* for userspace compile, we disable error debugging */
+#ifndef _KERNEL
+#define	printk(fmt, ...)
+#endif
+
+/* User space tests compatibility */
+#ifndef _KERNEL
+#define	__init
+#define	__exit
+#endif
+
+/* These enums are index references to zstd_cache_config */
+enum zstd_kmem_type {
+	ZSTD_KMEM_UNKNOWN = 0,
+	ZSTD_KMEM_DEFAULT,
+	ZSTD_KMEM_POOL,
+	ZSTD_KMEM_DCTX,
+	ZSTD_KMEM_COUNT,
+};
+
+struct zstd_pool {
+	void *mem;
+	size_t size;
+	kmutex_t 		barrier;
+	time_t timeout;
+};
+
+struct zstd_kmem {
+	enum zstd_kmem_type	kmem_type;
+	size_t			kmem_size;
+	struct zstd_pool	*pool;
+};
+
+struct zstd_fallback_mem {
+	size_t			mem_size;
+	void			*mem;
+	kmutex_t 		barrier;
+};
+
+struct levelmap {
+	int32_t cookie;
+	enum zio_zstd_levels level;
+};
+
+static void *zstd_alloc(void *opaque, size_t size);
+static void *zstd_dctx_alloc(void *opaque, size_t size);
+static void zstd_free(void *opaque, void *ptr);
+
+static const ZSTD_customMem zstd_malloc = {
+	zstd_alloc,
+	zstd_free,
+	NULL,
+};
+
+static const ZSTD_customMem zstd_dctx_malloc = {
+	zstd_dctx_alloc,
+	zstd_free,
+	NULL,
+};
```

Comment:
All those pool const, strucst etc could use a little commenting

Reply:
With the bit of added commenting in #8941 I think devs should be able to understand the rest themselves rather quickly now. Basically (Re)Solved in #8941